### ISSUE TITLE
DDP - pytorch多机多卡训练

### ISSUE THREAD

**Issue by iLovEing** [iLovEing/notebook#32]
# torch.nn.DistributedDataParallel

**Comment by iLovEing** [comment]
2.3 cuda()函数解释
cuda() 函数返回一个存储在CUDA内存中的复制，其中device可以指定cuda设备。 但如果此storage对象早已在CUDA内存中存储，并且其所在的设备编号与cuda()函数传入的device参数一致，则不会发生复制操作，返回原对象。

cuda()函数的参数信息:

device ([int](https://docs.python.org/3/library/functions.html#int)) – 指定的GPU设备id. 默认为当前设备，即 [torch.cuda.current_device()](https://www.cntofu.com/book/169/docs/1.0/cuda.html#torch.cuda.current_device)的返回值。

non_blocking ([bool](https://docs.python.org/3/library/functions.html#bool)) – 如果此参数被设置为True, 并且此对象的资源存储在固定内存上(pinned memory)，那么此cuda()函数产生的复制将与host端的原storage对象保持同步。否则此参数不起作用。

**Comment by iLovEing** [comment]
# DDP 基本概念
- **rank**：进程，我们通常假定rank 0是第一个进程或者主进程，其它进程分别具有1，2，3不同rank号，这样总共具有4个进程

**Comment by iLovEing** [comment]
# DDP 基本流程 & 代码

### 训练流程

使用DDP分布式训练，一共如下几个步骤：

1. 初始化进程组 dist.init_process_group
2. 设置分布式采样器 DistributedSampler
3. 使用DistributedDataParallel封装模型
4. 使用torchrun 或者 mp.spawn 启动分布式训练

### 代码

torchrun 启动方式方便一些，这里只展示torchrun代码。通过一个MNIST训练实例来看ddp如何实现，方便起见，train和eval用相同的数据。

- **without DDP**
  保存代码为 mnist.py，命令行输入 python mnist.py 可直接运行
  ```
  import os
  import argparse
  from datetime import datetime
  import torch
  import torch.nn as nn
  import torchvision
  from tqdm import tqdm
  import torchvision.transforms as transforms
  from torch.utils.data import Dataset, DataLoader
  
  
  class ConvNet(nn.Module):
      def __init__(self, num_classes=10):
          super(ConvNet, self).__init__()
          self.layer1 = nn.Sequential(
              nn.Conv2d(1, 16, kernel_size=5, stride=1, padding=2),
              nn.BatchNorm2d(16),
              nn.ReLU(),
              nn.MaxPool2d(kernel_size=2, stride=2))
          self.layer2 = nn.Sequential(
              nn.Conv2d(16, 32, kernel_size=5, stride=1, padding=2),
              nn.BatchNorm2d(32),
              nn.ReLU(),
              nn.MaxPool2d(kernel_size=2, stride=2))
          self.fc = nn.Linear(7 * 7 * 32, num_classes)
  
      def forward(self, x):
          out = self.layer1(x)
          out = self.layer2(out)
          out = out.reshape(out.size(0), -1)
          out = self.fc(out)
          return out
  
  
  def train(args):
      # 0. preset
      batch_size = args.batch_size
      epochs = args.epochs
  
      device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
  
      # 1. model
      model = ConvNet()
      model.to(device)
  
      # 2. dataset
      train_set = torchvision.datasets.MNIST(root='./data', train=True,
                                             transform=transforms.ToTensor(), download=True)
      train_loader = DataLoader(dataset=train_set, batch_size=batch_size,
                                shuffle=True, pin_memory=True)
      eval_set = torchvision.datasets.MNIST(root='./data', train=False,
                                            transform=transforms.ToTensor(), download=True)
      eval_loader = DataLoader(dataset=eval_set, batch_size=batch_size,
                               shuffle=True, pin_memory=True)
  
      # 3. loss and optimizer
      criterion = nn.CrossEntropyLoss().to(device)
      optimizer = torch.optim.SGD(model.parameters(), 1e-4)
  
      # 4. training & eval
      start = datetime.now()
      step = 0
      for epoch in range(epochs):
          # train
          model.train()
          progress_bar_train = tqdm(train_loader, desc=f'Epoch {epoch + 1}/{epochs} train')
          for batch in progress_bar_train:
              image, label = batch
              image = image.to(device, non_blocking=True)
              label = label.to(device, non_blocking=True)
              # Forward pass
              outputs = model(image)
              loss = criterion(outputs, label)
  
              # Backward and optimize
              optimizer.zero_grad()
              loss.backward()
              optimizer.step()
              if step % 10 == 0:
                  progress_bar_train.set_postfix(loss=loss.item())
              step += 1
  
          with torch.no_grad():
              model.eval()
              preds = []
              targets = []
              progress_bar_eval = tqdm(eval_loader, desc=f'Epoch {epoch + 1}/{epochs} eval')
              for batch in progress_bar_eval:
                  image, label = batch
                  image = image.to(device, non_blocking=True)
                  label = label.to(device, non_blocking=True)
                  output = model(image)
                  pred = output.argmax(dim=1)
  
                  preds.append(pred)
                  targets.append(label)
              preds = torch.cat(preds)
              targets = torch.cat(targets)
              acc = (preds == targets).sum().cpu().numpy()
              print(f'Epoch {epoch + 1}/{epochs} acc: {format(acc, ".5f")}')
  
      print("Training complete in: " + str(datetime.now() - start))
      # 5. save
      torch.save(model.state_dict(), 'mnist.pth')
  
  
  def main():
      parser = argparse.ArgumentParser()
      parser.add_argument('--epochs', default=10, type=int)
      parser.add_argument('--batch_size', default=128, type=int)
      args = parser.parse_args()
      train(args)
  
  
  # run: python mnist.py
  if __name__ == '__main__':
      main()
  ```

- **add DDP code (torchrun)**
保存代码为 ddp_mnist.py，使用torch run运行，其中nnodes为机器数量，nproc_per_node为每台机器的gpu数量，如果在多机多卡上运行，对应机器的启动命令中需要修改node_rank为机器的编号：
torchrun --nnodes=1 --node_rank=0 --nproc_per_node=2 --master_addr="192.168.1.250" --master_port=23456 ddp_mnist.py
```
import os
import argparse
from datetime import datetime
import torch
import torch.nn as nn
import torchvision
from tqdm import tqdm
import torchvision.transforms as transforms
from torch.utils.data import Dataset, DataLoader
########## DDP ##########
import torch.distributed as dist
from torch.utils.data.distributed import DistributedSampler
###########################


class ConvNet(nn.Module):
    def __init__(self, num_classes=10):
        super(ConvNet, self).__init__()
        self.layer1 = nn.Sequential(
            nn.Conv2d(1, 16, kernel_size=5, stride=1, padding=2),
            nn.BatchNorm2d(16),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2))
        self.layer2 = nn.Sequential(
            nn.Conv2d(16, 32, kernel_size=5, stride=1, padding=2),
            nn.BatchNorm2d(32),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2))
        self.fc = nn.Linear(7 * 7 * 32, num_classes)

    def forward(self, x):
        out = self.layer1(x)
        out = self.layer2(out)
        out = out.reshape(out.size(0), -1)
        out = self.fc(out)
        return out


########## DDP-6 ##########
@torch.no_grad()
def concat_all_gather(tensor):
    """
    Performs all_gather operation on the provided tensors.
    *** Warning ***: torch.distributed.all_gather has no gradient.
    """
    tensors_gather = [torch.ones_like(tensor) for _ in range(torch.distributed.get_world_size())]
    torch.distributed.all_gather(tensors_gather, tensor, async_op=False)

    output = torch.cat(tensors_gather, dim=0)
    return output
###########################


def train(args):
    # 0. preset
    batch_size = args.batch_size
    epochs = args.epochs

    ########## DDP-1 ##########
    # device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
    gpu = int(os.environ["LOCAL_RANK"])
    rank = int(os.environ["RANK"])
    world_size = int(os.environ['WORLD_SIZE'])
    dist.init_process_group(
        backend='nccl',
        init_method='env://',
        world_size=world_size,
        rank=rank
    )
    device = torch.device(f"cuda:{gpu}")
    print(f'pid: {os.getpid()}, ppid: {os.getppid()}, gpu: {gpu}-{rank}')
    ###########################

    # 1. model
    model = ConvNet()
    model.to(device)
    ########## DDP-2 ##########
    model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)  # for batch normal layer
    model = nn.parallel.DistributedDataParallel(model, device_ids=[gpu])
    ###########################

    # 2. dataset
    ########## DDP-3 ##########
    train_set = torchvision.datasets.MNIST(root='./data', train=True,
                                           transform=transforms.ToTensor(), download=True)
    # train_loader = DataLoader(dataset=train_set, batch_size=batch_size,
    #                           shuffle=True, pin_memory=True)
    # dataset will be divided into {world_size} parts automatic by sampler,
    # every rank will obtain its own dataset
    train_sampler = DistributedSampler(train_set, num_replicas=world_size, rank=rank)
    train_loader = DataLoader(train_set,
                              batch_size=batch_size,
                              shuffle=False,  # sampler will do shuffle
                              sampler=train_sampler)

    eval_set = torchvision.datasets.MNIST(root='./data', train=False,
                                          transform=transforms.ToTensor(), download=True)
    # eval_loader = DataLoader(dataset=eval_set, batch_size=batch_size,
    #                          shuffle=True, pin_memory=True)
    eval_sampler = DistributedSampler(eval_set, num_replicas=world_size, rank=rank)
    eval_loader = DataLoader(eval_set, batch_size=batch_size, shuffle=False, sampler=eval_sampler)
    #########################

    # 3. loss and optimizer
    criterion = nn.CrossEntropyLoss().to(device)
    optimizer = torch.optim.SGD(model.parameters(), 1e-4)

    # 4. training & eval
    start = datetime.now()
    step = 0
    for epoch in range(epochs):
        ########## DDP-4 ##########
        train_sampler.set_epoch(epoch)
        eval_sampler.set_epoch(epoch)
        ###########################

        # train
        model.train()
        ########## DDP-5 ##########
        # progress_bar_train = tqdm(train_loader, desc=f'Epoch {epoch + 1}/{epochs} train')
        progress_bar_train = tqdm(train_loader, desc=f'Epoch {epoch + 1}/{epochs} train') \
            if gpu == 0 else train_loader
        ##########################
        for batch in progress_bar_train:
            image, label = batch
            image = image.to(device, non_blocking=True)
            label = label.to(device, non_blocking=True)
            # Forward pass
            outputs = model(image)
            loss = criterion(outputs, label)

            # Backward and optimize
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            ########## DDP-5 ##########
            # if step % 10 == 0:
            if gpu == 0 and step % 10 == 0:
                progress_bar_train.set_postfix(loss=loss.item())
            ##########################
            step += 1

        with torch.no_grad():
            model.eval()
            preds = []
            targets = []
            ########## DDP-5 ##########
            # progress_bar_eval = tqdm(eval_loader, desc=f'Epoch {epoch + 1}/{epochs} eval')
            progress_bar_eval = tqdm(eval_loader, desc=f'Epoch {epoch + 1}/{epochs} eval') \
                if gpu == 0 else eval_loader
            ##########################
            for batch in progress_bar_eval:
                image, label = batch
                image = image.to(device, non_blocking=True)
                label = label.to(device, non_blocking=True)
                output = model(image)
                pred = output.argmax(dim=1)

                ########## DDP-6 ##########
                pred = concat_all_gather(pred)
                label = concat_all_gather(label)
                ##########################
                preds.append(pred)
                targets.append(label)

            ########## DDP-5 ##########
            if gpu == 0:
                preds = torch.cat(preds)
                targets = torch.cat(targets)
                acc = (preds == targets).sum().cpu().numpy() / len(preds)
                print(f'Epoch {epoch + 1}/{epochs} acc: {format(acc, ".5f")}')
            ##########################

    ########## DDP-5 ##########
    if gpu == 0:
        print("Training complete in: " + str(datetime.now() - start))
    ##########################
    # 5. save
    if rank == 0:
        torch.save(model.state_dict(), 'mnist.pth')
    ##########################


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--epochs', default=10, type=int)
    parser.add_argument('--batch_size', default=128, type=int)
    args = parser.parse_args()
    train(args)

########## DDP-0 ##########
# run: torchrun --nnodes=1 --node_rank=0 --nproc_per_node=2 --master_addr="192.168.1.250" --master_port=23456 ddp_mnist.py
##########################
if __name__ == '__main__':
    main()
```

**Comment by iLovEing** [comment]
# 代码&api 解释

 ### 1. DDP-0：启动分布式训练

使用torchrun方式启动，会自动设置进程的RANK,LOCAL_RANK, WORLD_SIZE等，可以直接从环境变量获取。常用的参数有：

- `--nnodes`: 使用的机器数量，单机的话，默认是1
- `--nproc_per_node`: 每台机器的进程数，通常为每台机器参与训练的GPU数
- `--master_addr/port`: 主进程rank0所在的机器地址和后端通信端口（端口不被占用即可）
- `--node_rank`: 当前机器的id

在实例代码中，就是单机多卡的启动示例。如果使用单机多卡，只有`nproc_per_node` 是必须指定的，`master_addr/port`和`node_rank`都是可以由launch通过环境自动配置；如果使用多机多卡，不同机器都要输入启动命令，其中每台机器的`--node_rank`参数需要修改为本机id，须保证`node_rank < nnodes`，其他参数相同。

### 2. DDP-1：初始化进程组

使用`torch.distributed.init_process_group`初始化进程组，参数：

- `backend`： 使用的后端

- `init_method`：初始化方法，主要是指获取通信地址、端口、rank、world size的方式。

  - `init_method='tcp://ip:port`：通过指定rank 0（即：MASTER进程）的IP和端口，各个进程进行信息交换。 需指定 rank 和 world_size 这两个参数
  - `init_method='file://path`：通过所有进程都可以访问共享文件系统来进行信息共享。需要指定rank和world_size参数
  - `init_method=env://`：从环境变量中读取分布式的信息(os.environ)，主要包括 MASTER_ADDR, MASTER_PORT, RANK, WORLD_SIZE。 其中，rank和world_size可以选择手动指定，否则从环境变量读取 

  tcp和env两种方式比较类似（其实env就是对tcp的一层封装），都是通过网络地址的方式进行通信，也是最常用的初始化方法

- `world_size`：即world size

- `rank`：当前进程的rank

### 3. DDP-2：封装模型

直接用`torch.nn.parallel.DistributedDataParallel`封装即可，封装以后，框架会自动做不同GPU上梯度的all gather；如果有batch normal layer，想要做全局的batch normal，则需要用`convert_sync_batchnorm`单独封装。

### 4. DDP-3：封装dataset

使用`torch.utils.data.distributed.DistributedSampler`封装dataloader，这样dataset会自动划分为word size份，并送到每个进程进行训练，这里shuffle置为False，因为DistributedSampler已经做了打乱操作。同时，需要在每个epoch开始前对sampler进行set epoch，以保证数据打乱并进行不同的划分。`DistributedSampler`部分参数：

- `dataset`: 需要加载的完整数据集
- `num_replicas`： 把数据集分成多少份，默认是当前dist的world_size
- `rank`: 当前进程的id，默认dist的rank
- `shuffle`：是否打乱
- `drop_last`: 如果数据长度不能被world_size整除，可以考虑是否将剩下的扔掉
- `seed`：随机数种子。这里需要注意，从源码中可以看出，真正的种子其实是 `self.seed+self.epoch` 这样的好处是，不同的epoch每个进程拿到的数据是不一样

### 5. DDP-4：多进程相关操作

观察`DDP-4`相关的代码，会发现这里大都是指定rank或者local_rank进行的动作，这是因为ddp是以多进程方式执行，这里每行代码都会在各自的进程运行，因此需要开发者仔细考虑哪些部分是只要在主进程内运行的，比如希望log在每个节点上打印一次（不重复打印），则指定local_rank为0的进程运行；希望只在主节点保存一份模型，则指定rank为0的进程运行。

### 6. DDP-5：分布式推理

分布式推理的基本处理方式和训练一样，唯一不同的是需要自己做数据整合，即把多卡上的目标数据进行整合合并。常用的两个函数：

- `torch.distributed.all_gather`： 将所有进程的tensor进行收集并拼接成新的tensor list内
- `torch.distributed.all_reduce`：对所有进程的某个tensor进行合并操作（in-place），op可以是求和等，返回op的结果

**Comment by iLovEing** [comment]
# 其他

### 1. 常用api

- `torch.distributed.get_rank(group=None)`： 获取rank
- `torch.distributed.get_backend(group=None)`： 获取当前任务（或者指定group）的后端
- `torch.distributed.get_world_size(group=None)` 获取当前进程的rank
- `dist.barrier()`：类似多进程的join，所有进程都运行到此处时才往下运行，否则等待

### 2. 关于cuda device

首先，进程可见的cuda设备取决于`os.environ["CUDA_VISIBLE_DEVICES"]`的值，假如一台机器上有5块GPU，驱动会对其有内置的编号0~4，如果设置了`os.environ["CUDA_VISIBLE_DEVICES"]=“3, 1, 4, 0”`，那么此时运行环境的gpu编号对应关系为：

> cuda:0 --> 3
>
> cuda:1 --> 1
>
> cuda:2 --> 4
>
> cuda:3 --> 0

DDP中local rank也是一样。

第二，示例代码中使用`to(device)`指定GPU，也可以使用`.cuda()`。cuda() 函数返回一个存储在CUDA内存中的复制，其中device可以指定cuda设备。 但如果此storage对象早已在CUDA内存中存储，并且其所在的设备编号与cuda()函数传入的device参数一致，则不会发生复制操作，返回原对象。函数参数：

- device：指定的GPU设备id。 默认为当前设备，即 `torch.cuda.current_device()`的返回值，可以通过`torch.cuda.set_device(gpu)`提前修改默认值。

- non_blocking ：如果此参数被设置为True, 并且此对象的资源存储在固定内存上(pinned memory)，那么此cuda()函数产生的复制将与host端的原storage对象保持同步。否则此参数不起作用。

### 3. small tips

- 由于model在ddp中封装了一层，因此在save model时需要保存`model.module.state_dict()`，否则也不会报错，但是等load的时候，你就知道哭了。
- DDP中，batch size是指每个节点(GPU)的batch，因此实际的batch size应为 `batch_size * world_slze`
